JPA N+1 쿼리 문제 해결을 위한 리팩토링

kimcno3 opened the pull request:
* ChatServiceImpl
  - sendMessage() 로직 흐름 설명을 위한 주석문 추가
  - N+1쿼리문제가 아닌 지연로딩 문제인지 아닌지에 대한 고민을 적은 주석문 추가

kimcno3 committed:
JPA N+1쿼리 리팩토링

* ChatServiceImpl
  - sendMessage() 로직 흐름 설명을 위한 주석문 추가
  - N+1쿼리문제가 아닌 지연로딩 문제인지 아닌지에 대한 고민을 적은 주석문 추가

kimcno3 commented on src/main/java/com/flab/football/service/chat/ChatServiceImpl.java:
```diff
@@ -162,6 +157,9 @@ public List<Integer> findMessageReceivers(int channelId) {
+      // participant 객체에 저장된 User 객체를 사용하기 위해 N번의 쿼리문 발생
+
+      // 잠깐! N+1 쿼리라기 보다는 연관관계를 가진 객체를 조회하는 지점에서 쿼리문이 발생하는 지연 로딩으로 인한 N번의 쿼리문 발생 지점이진 않은지...
       userIdList.add(participant.getUser().getId());
```
멘토님 이전 리뷰에서 N+1 쿼리가 발생하는 지점으로 보였던 코드가 N+1이 실제로 발생하고 있는지에 대한 의문이 들었습니다!

APM으로 확인해보면 해당 코드가 실행되는 시점에서 N개의 participant 객체와 연관관계 객체인 user가 조회되는 시점에서 지연로딩 되었던 쿼리문이 반복문으로 인해 N번 발생한 것이 아닌지 궁금합니다!!

kimcno3 replied:
![스크린샷 2022-08-17 오전 2 27 11](https://user-images.githubusercontent.com/77563468/184941979-7566d465-a45a-4c87-8f22-412b5d3b7dab.png)

![스크린샷 2022-08-17 오전 2 27 25](https://user-images.githubusercontent.com/77563468/184942031-d05dcf31-9a9e-4e92-b24c-06039b5d6dfc.png)

이미지는 APM으로 나온 sendMessage() 의 trace 모습입니다!
findParticipantsByChannelId()가 실행되는 시점에서 반복문의 길이만큼 쿼리문이 발생했습니다!

f-lab-bright replied:
얼라 근데 마지막에 message update 쿼리는 왜 발생하나요..?
메세지에 대한 insert만 일어나면 되지 않나요??

kimcno3 replied:
아마 수정 전 코드가 message 객체를 먼저 save한 한 다음 연관관계 객체를 setter로 추가해 update 쿼리가 발생한 것으로 예상됩니다!

이후에는 setter로 연관관계 객체까지 추가한 뒤에 save()로 insert 쿼리를 호출할 수 있도록 구성해 update문은 발생하지 않습니다!

kimcno3 replied:
```
Hibernate: select channel0_.id as id1_0_0_, channel0_.name as name2_0_0_ from channel channel0_ where channel0_.id=?
Hibernate: select user0_.id as id1_8_0_, user0_.email as email2_8_0_, user0_.gender as gender3_8_0_, user0_.name as name4_8_0_, user0_.password as password5_8_0_, user0_.phone as phone6_8_0_, user0_.role as role7_8_0_ from user user0_ where user0_.id=?
Hibernate: insert into message (channel_id, content, create_at, type, user_id) values (?, ?, ?, ?, ?)
Hibernate: select participan0_.id as id1_5_0_, user1_.id as id1_8_1_, participan0_.channel_id as channel_2_5_0_, participan0_.user_id as user_id3_5_0_, user1_.email as email2_8_1_, user1_.gender as gender3_8_1_, user1_.name as name4_8_1_, user1_.password as password5_8_1_, user1_.phone as phone6_8_1_, user1_.role as role7_8_1_ from participant participan0_ left outer join user user1_ on participan0_.user_id=user1_.id where participan0_.channel_id=?
```

수정 후 메세지 전송 로직 실행시 발생하는 쿼리문입니다!

kimcno3 commented on src/main/java/com/flab/football/service/chat/ChatServiceImpl.java:
```diff
@@ -100,15 +102,16 @@ public void sendMessage(int channelId, int sendUserId, String content) {
     message.setUser(user);
 
+    // channel 정보 주입
     message.setChannel(channel);
```
그리고 해당 로직처럼 messge 객체에 모든 연관관계 객체를 주입해주는 로직이 먼저 수행되도록 작성되었지만 실제 전송된 쿼리문 순서를 보면 가장 마지막에 실행되는 것을 볼 수 있습니다.

이 부분은 JPA 교재를 다시 보면서 개념을 익혀볼 필요가 있을 지 궁금합니다!!

```
1) channel 객체 조회
Hibernate: select channel0_.id as id1_0_0_, channel0_.name as name2_0_0_ from channel channel0_ where channel0_.id=?

2) user 객체 조회
Hibernate: select user0_.id as id1_8_0_, user0_.email as email2_8_0_, user0_.gender as gender3_8_0_, user0_.name as name4_8_0_, user0_.password as password5_8_0_, user0_.phone as phone6_8_0_, user0_.role as role7_8_0_ from user user0_ where user0_.id=?

3) message 객체 저장
Hibernate: insert into message (channel_id, content, create_at, type, user_id) values (?, ?, ?, ?, ?)

4) channelId에 따른 모든 participant 객체 조회
Hibernate: select participan0_.id as id1_5_, participan0_.channel_id as channel_2_5_, participan0_.user_id as user_id3_5_ from participant participan0_ left outer join channel channel1_ on participan0_.channel_id=channel1_.id where channel1_.id=?

5) 각 participant 객체에 대한 user 객체 조회 -> 이후 Spring에서 userId 조회
Hibernate: select user0_.id as id1_8_0_, user0_.email as email2_8_0_, user0_.gender as gender3_8_0_, user0_.name as name4_8_0_, user0_.password as password5_8_0_, user0_.phone as phone6_8_0_, user0_.role as role7_8_0_ from user user0_ where user0_.id=?
Hibernate: select user0_.id as id1_8_0_, user0_.email as email2_8_0_, user0_.gender as gender3_8_0_, user0_.name as name4_8_0_, user0_.password as password5_8_0_, user0_.phone as phone6_8_0_, user0_.role as role7_8_0_ from user user0_ where user0_.id=?
Hibernate: select user0_.id as id1_8_0_, user0_.email as email2_8_0_, user0_.gender as gender3_8_0_, user0_.name as name4_8_0_, user0_.password as password5_8_0_, user0_.phone as phone6_8_0_, user0_.role as role7_8_0_ from user user0_ where user0_.id=?
Hibernate: select user0_.id as id1_8_0_, user0_.email as email2_8_0_, user0_.gender as gender3_8_0_, user0_.name as name4_8_0_, user0_.password as password5_8_0_, user0_.phone as phone6_8_0_, user0_.role as role7_8_0_ from user user0_ where user0_.id=?

6) message 객체에 channelId 와 userId 값을 update -> 먼저 작성된 로직과는 다르게 가장 마지막에 수행되는 쿼리문
Hibernate: update message set channel_id=?, content=?, create_at=?, type=?, user_id=? where id=?
```

f-lab-bright replied:
jpa 에서 lazy write 가 어떻게 동작하는지 확인해보시는게 필요할 것 같습니다 :)
lazy write로 동작하기 위해 jpa는 query buffer에 쌓인 쿼리를 특정한 순서에 따라 실행합니다.
이로 인해 의도하지 않은 순서로 쿼리가 실행되기도 하는데, 이런 상황이 발생할다면, 특정 시점에 직접 flush 하는 등으로 해결해야 할 수도 있습니다 :)

f-lab-bright commented on src/main/java/com/flab/football/service/chat/ChatServiceImpl.java:
```diff
@@ -162,6 +157,9 @@ public List<Integer> findMessageReceivers(int channelId) {
+      // participant 객체에 저장된 User 객체를 사용하기 위해 N번의 쿼리문 발생
+
+      // 잠깐! N+1 쿼리라기 보다는 연관관계를 가진 객체를 조회하는 지점에서 쿼리문이 발생하는 지연 로딩으로 인한 N번의 쿼리문 발생 지점이진 않은지...
       userIdList.add(participant.getUser().getId());
```
이 요청이 처리될 때 apm으로 쿼리가 어떻게 실행되고 있는지 확인해보셨나요??
어떤 쿼리들이 실행되나요??
채팅방에 10명 정도가 있는 상황일 때 어떤 쿼리들이 몇 건이 발생하나요??

kimcno3 replied:
해당 코드에서 발생하는 쿼리는 다음과 같습니다!

```
select user0_.id as id1_8_0_, user0_.email as email2_8_0_, user0_.gender as gender3_8_0_, user0_.name as name4_8_0_, user0_.password as password5_8_0_, user0_.phone as phone6_8_0_, user0_.role as role7_8_0_ from user user0_ where user0_.id=?
```

그리고 채팅방 인원을 10명으로 추가한 다음 다시 요청을 보내보니 10건의 위 쿼리가 발생한 것을 확인했습니다!!

<img width="1145" alt="스크린샷 2022-08-19 오전 2 39 05" src="https://user-images.githubusercontent.com/77563468/185459533-49b6477e-422a-4067-bad0-9fc5b34ec0b6.png">

kimcno3 replied:
제가 보기엔 해당 쿼리는 연관관계 객체를 조회하는 시점에서 지연 로딩으로 인해 발생하는 쿼리문이 participant 객체별로 발생하는 것이라고 판단하고 있습니다! 그래서 해당 부분을 N+1 쿼리로 보는 것이 적합할지에 대한 의문이 있습니다ㅠㅠ

f-lab-bright commented:
n+1 쿼리가 jpa에서의 구조적인 문제로 쿼리 조회시 자동으로 발생하는 경우도 문제지만, 로지컬리하게 직접 다량의 쿼리를 유사하게 발생시키는 경우도 문제입니다.
jpa의 특성으로 발생하는 n+1 쿼리는 아니지만 현재 유즈케이스에서는 선호님이 작성한 코드가 직접 호출하는 코드로 인해서 n+1 문제를 만들고 있는겁니다.
이것도 당연히 해결해야 하는 코드입니다 :)

kimcno3 commented:
네 알겠습니다!! N+1 쿼리라고 보기 애매한 현상이라면 다른 제목으로 해결해보고자 했었는데 그대로 N+1 쿼리로 보고 해결해보도록 하겠습니다 😁

kimcno3 committed:
JPA N+1쿼리 리팩토링

- ChatServiceImpl.sendMessage() 로직 흐름 설명을 위한 주석문 추가
- N+1쿼리문제가 아닌 지연로딩 문제인지 아닌지에 대한 고민을 적은 주석문 추가
- 연관관계 엔티티 필드에 대해 cascadeType.ALL 추가
- 객체 생성 및 저장 메소드에 불필요한 save() 호출 코드 삭제

kimcno3 committed:
JPA N+1쿼리 리팩토링

- EntityGraph 어노테이션을 활용한 N+1 쿼리 발생 로직 제거

kimcno3 commented:
지금까지 아래 세가지 방법으로 N+1 쿼리 발생 문제를 해결해봤고 각 방법의 단점을 분석해보고 그 내용은 블로그 글로 정리해봤습니다!!

1. 단순하게 외래키를 하나의 필드로만 가져와 해결해보는 방법
2. Fetch Join JPQL을 직접 선언해주는 방법
3. EntityGraph를 활용해 Fetch Join 설정을 해주는 방법

결국 제가 선택한 방법은 3번 방법이고 그렇게 생각한 이유는

지금 당장에는 userId만을 호출하기 때문에 Fetch Join을 활용하는 것이 불필요한 영속성 컨텍스트에 불필요하게 메모리를 할당하는 것이라 볼 수도 있겠지만, 이후 추가적인 기능 구현으로 인해 연관관계로 엔티티 구조를 변경한다면, 서비스 레이어에 선언된 비즈니스 로직의 코드도 수정 소요가 발생하면서 확장에 불리한 구조가 될 것 같다는 판단을 했습니다!

정리한 블로그 글 [링크](https://velog.io/@kimcno3/N1-%EC%BF%BC%EB%A6%AC-%EB%AC%B8%EC%A0%9C-%ED%95%B4%EA%B2%B0%EC%97%90-%EB%8C%80%ED%95%9C-%EA%B3%A0%EB%AF%BC)입니다!

kimcno3 committed:
JPA N+1쿼리 리팩토링

- Channel - Message 일대다 연관관계로 인한 List 타입의 필드를 주석처리

kimcno3 commented:
추가로 이전에 멘토링 시간에 말씀해주셨던 Channel 엔티티 객체에서 Message 객체를 리스트 형태로 가지고 있는 경우, 이후 매우 큰 N+1 쿼리문제를 야기할 수 있다 해서 일단 주석문 처리를 해두었습니다!

f-lab-bright commented on src/main/java/com/flab/football/repository/chat/ParticipantRepository.java:
```diff
@@ -16,6 +18,8 @@ public interface ParticipantRepository extends JpaRepository<Participant, Intege
 
   Optional<Participant> findById(Participant participant);
 
+  @EntityGraph(attributePaths = "user")
+  @Query("SELECT p FROM Participant p")
```
이렇게 하면 특정 채널의 참여자가 아니라, 참여자 테이블의 모든 참여자가 조회되지 않을까요?

kimcno3 replied:
JPQL문에 channelId에 대한 where절을 추가했습니다! DB에 저장된 channel이 하나여서 제대로된 테스트 결과를 얻지 못했었네요ㅠㅠ

f-lab-bright commented on src/main/java/com/flab/football/domain/Channel.java:
```diff
@@ -38,8 +38,12 @@ public class Channel {
   @OneToMany(mappedBy = "channel", cascade = CascadeType.ALL)
   private List<Participant> participants;
 
+  /*
+   * 지속적으로 개수가 증가할 데이터인데 일대다 연관관계를 가지게 되면 심각한 N+1 쿼리를 발생할 수 있는 문제가 존재하지 않을까
+
```
N+1 문제보다, channel 조회 시 모든 메세지가 메모리에 로딩되는게 더 큰 문제입니다 :)
채널에서 많은 메세지가 발생하여 100만건 정도가 쌓여있다면, channel이 조회되고, message가 읽힌다면 100만건의 엔티티가 모두 메모리에 올라올텐데, 이는 OOM을 일으키기 딱 좋은 설계가 되겠죠 :)

kimcno3 replied:
넵! 말씀해주신 내용을 주석문에도 추가했습니다!

f-lab-bright commented on src/main/java/com/flab/football/service/chat/ChatServiceImpl.java:
```diff
@@ -98,17 +98,14 @@ public void sendMessage(int channelId, int sendUserId, String content) {
-    // N+1 쿼리 제어 필요
+    messageRepository.save(message);
+
     List<Integer> userIdList = findMessageReceivers(channelId);
```
결국 요 부분의 개선을 위해서 구조가 변경되거나, 쿼리를 최적화 해야 할 것 같은데, 이 부분은 아직 작업 전인건가요??

kimcno3 replied:
일단 위에서 수정한대로 N+1 쿼리 발생 부분을 Fetch Join으로 해결하면서 1차적으로 쿼리를 최적화 했다고 생각합니다!

그렇다면 이후에 추가로 더 최적화 해볼 만한 부분이 있을 지 궁금합니다!!

f-lab-bright replied:
userId만 가져오면 되는 상황에서 user table까지의 join이 굳이 필요할까요..?? :) 
user table의 join 없이 참여자의 userId 목록을 가져올 수 있는 방법도 있을까요..? :)

kimcno3 replied:
리팩토링한 내용을 푸시했고 지점마다 코멘트를 작성해봤습니다!!

kimcno3 committed:
findAllByChannelId() JPQL문 수정

* channelId에 대한 조건을 추가하기 위해 where 절 추가
* 매개변수 값을 JPQL문에 적용시키기 위해 @Param문 추가

kimcno3 committed:
주석문 추가

* Channel 에 message 객체 리스트를 받지 않아야 하는 이유에 대한 멘토님 의견을 주석문에 추가 작성

kimcno3 committed:
피드백 반영

* Participant와 연관관계를 가지는 객체들에 insertable, updatable 값을 false로 설정
  -> insert, update시 해당 연관관계 값은 대상에서 제외

* userId, channelId를 column으로 추가 선언

* ParticipantRepository.findAllUserIdByChannelId() 생성
  -> JPQL문을 직접 작성해 userId만 조회하도록 설정

* ChatServiceImpl.inviteParticipants() 로직 변경
  -> 연관관계 객체 조회 쿼리문을 발생시키는 메소드들 삭제
  -> 매개변수로 넘어온 id값을 컬럼값으로 선언만 해주도록 변경

* ChatServiceImpl.findMessageReceivers 로직 변경
  -> findAllUserIdByChannelId()로 userId만 조회하도록 수정
  -> 코드 수를 최소화

kimcno3 commented on src/main/java/com/flab/football/domain/Participant.java:
```diff
@@ -35,12 +35,20 @@ public class Participant {
   @OneToOne
-  @JoinColumn(name = "user_id", referencedColumnName = "id")
+  @JoinColumn(name = "user_id", referencedColumnName = "id",
+              insertable = false, updatable = false)
```
연관관계 객체들에 대해 insertable, updatable 설정을 false로 변경해주면서 insert, update 시 해당 객체들을 대상에서 제외했습니다.

kimcno3 commented on src/main/java/com/flab/football/domain/Participant.java:
```diff
@@ -35,12 +35,20 @@ public class Participant {
+
+  @Column(name = "user_id")
+  private int userId;
+
```
insertable, updatable 설정을 false로 하면서 직접 외래키를 하나의 필드값으로 선언할 수 있게 되었습니다!

f-lab-bright replied:
유즈케이스에 따라 다를 수 있겠지만, Participant를 통해서 User를 수정할 일이 있다면, 반대로 해두시는게 좋을 수도 있습니다 :)
그렇지 않다면 지금 처럼 하면 될 것 같고, updatable은 여기도 false로 해두는게 안전할 수 있겠네요 :)

kimcno3 commented on src/main/java/com/flab/football/repository/chat/ParticipantRepository.java:
```diff
@@ -23,4 +23,7 @@ public interface ParticipantRepository extends JpaRepository<Participant, Intege
 
+  @Query("SELECT p.userId FROM Participant p WHERE p.channel.id = :channelId")
+  List<Integer> findAllUserIdByChannelId(@Param(value = "channelId") int channelId);
+
```
userId를 필드값으로 설정해두었기 때문에 굳이 fetch Join을 활용하지 않고도 userId만을 조회해오는 메소드를 선언할 수 있었습니다! 다만 직접 JPQL문을 작성해 @Query로 선언해두었습니다.

kimcno3 commented on src/main/java/com/flab/football/service/chat/ChatServiceImpl.java:
```diff
@@ -62,20 +61,13 @@ public void inviteParticipants(int channelId, List<Integer> participants) {
-      participant.setChannel(channel);
-
-      channel.addParticipant(participant);
-
```
그리고 위와 같은 구조로 리팩토링 하니 participant 객체를 생성하는 로직에서도 연관관계 객체를 쿼리문으로 조회해오는 과정을 생략할 수 있게 되면서 연관관계 구조로 인해 불필요하게 생성되는 쿼리문 발생 지점을 제거할 수 있는 것 같습니다!

f-lab-bright reviewed (approved):
LGTM
